src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp

```diff
@@ -0,0 +1,126 @@
+/**
+ * @file methods/xgboost/loss_functions/sse_loss.hpp
+ * @author Rishabh Garg
+ *
+ * The sum of squared error loss class, which is a loss funtion for gradient
+ * xgboost based decision trees.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_XGBOOST_LOSS_FUNCTIONS_SSE_LOSS_HPP
+#define MLPACK_METHODS_XGBOOST_LOSS_FUNCTIONS_SSE_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ensemble {
+
+/**
+ * The SSE (Sum of Squared Errors) loss is a loss function to measure the
+ * quality of prediction of response values present in the node of each
+ * xgboost tree. It is also a good measure to compare the spread of two
+ * distributions. We will try to minimize this value while training.
+ *
+ * Loss = 1 / 2 * (Observed - Predicted)^2
+ */
+class SSELoss
+{
+ public:
+  /**
+   * Returns the initial predition for gradient boosting.
+   */
+  template<typename VecType>
+  typename VecType::elem_type InitialPrediction(const VecType& values)
+  {
+    return arma::accu(values) / (typename VecType::elem_type) values.n_elem;
+  }
+
+  /**
+   * Returns the first order gradient of the loss function with respect to the
+   * values.
+   *
+   * This is primarily used in calculating the residuals and split gain for the
+   * gradient boosted trees.
+   *
+   * @tparam T The type of input data. This can be both a vector or a scalar.
+   * @param observed The true observed values.
+   * @param values The values with respect to which the gradient will be
+   *     calculated.
+   */
+  template<typename T>
+  T Gradients(const T& observed, const T& values)
+  {
+    return - (observed - values);
+  }
+
+  /**
+   * Returns the second order gradient of the loss function with respect to the
+   * values. This is used only for scalars.
+   */
+  template<typename T>
+  T Hessians(const T& /* observed */, const T& /* values */)
+  {
+    return (T) 1;
+  }
```

Comment:
Do we need an overload for scalars?  The reason that I ask is that it seems like `SSELoss` is implementing a set of methods that could be implemented by a different type of loss too.  That is, `SSELoss` is a template policy meant to be used with a gradient boosting class, and so if a user (or a maintainer) wanted to implement a new type of loss, they would need to implement the same functions that `SSELoss` implements.  That process is much easier when we can reduce the number of functions that need to be implemented to a minimum; so, if a scalar `Hessians()` overload isn't *really* needed and the result could be computed using the regular `Hessians()` overload, I might suggest using that one instead.

But, I have not seen the way that you are intending to use this function, so my comment may be misguided. :)

Reply:
I completely agree with your point. To be honest, I myself do not have absolute clarity about what functions we will be needing or not. I *think* that we might be needing this overload, so I thought I will directly use it when I will need it. But I might be wrong and we might not need it at all.

There are many things that are specific to loss functions and tasks that they can do. It's a little bit fuzzy area. This might become clearer when we are doing the splitting strategy.

I think maybe we can remove the unnecessary functions later?

Reply:
Sure, that sounds good, we just need to remember to go back once everything is working and think about whether we chose a good way to do it, or if the abstraction could be improved. :)  (But, you are right, it is very hard to predict the exact right interface to provide, and we may not get it right the first time.  That's okay!)